### PR TITLE
Improve package scanner version matching

### PR DESCRIPTION
**Pull request by sebasfalcone**
|Related issue|
|---|
| #21943 |

## Description
This PR adds the baseline for the improved logic for the package scanner version matching algorithms.

Changelog:
- Expanded package scanner version matcher map to cover all supported package formats:
  - Pypi -> PEP440
  - npm -> semVer
- Added new matching strategies for packages
 - MacOS -> DPKG
 - Windows -> DPKG
 - Pacman, Snap, PKG -> Try all version matchers

## Tests
Compared new results against 4.8.0:

### Diffs
New vulnerabilities are detected:

Package | version | CVEid
-- | -- | --
pip | 22.0.2 | CVE-2023-5752
setuptools | 59.6.0 | CVE-2022-40897
setuptools | 56.0.0 | CVE-2022-40897
setuptools | 56.0.0 | CVE-2022-40897
Pygments | 2.11.2 | CVE-2022-40896

<details>
<summary><b>Pygments</b></summary>

### 4.8.0
- Unaffected (CVE-2015-8557, CVE-2021-20270, CVE-2021-27291,  CVE-2022-40896)
```
2024/03/06 09:28:37 wazuh-modulesd:vulnerability-scanner[58809] packageScanner.hpp:271 at operator()(): DEBUG: No match due to default status for Package: pygments, Version: 2.16.1 while scanning for Vulnerability: CVE-2015-8557
2024/03/06 09:28:37 wazuh-modulesd:vulnerability-scanner[58809] packageScanner.hpp:271 at operator()(): DEBUG: No match due to default status for Package: pygments, Version: 2.16.1 while scanning for Vulnerability: CVE-2021-20270
2024/03/06 09:28:37 wazuh-modulesd:vulnerability-scanner[58809] packageScanner.hpp:271 at operator()(): DEBUG: No match due to default status for Package: pygments, Version: 2.16.1 while scanning for Vulnerability: CVE-2021-27291
2024/03/06 09:28:37 wazuh-modulesd:vulnerability-scanner[58809] packageScanner.hpp:271 at operator()(): DEBUG: No match due to default status for Package: pygments, Version: 2.16.1 while scanning for Vulnerability: CVE-2022-40896

```

### New branch
- Unaffected (CVE-2015-8557, CVE-2021-20270, CVE-2021-27291,  CVE-2022-40896)
```
2024/03/05 23:39:44 wazuh-modulesd:vulnerability-scanner[394838] packageScanner.hpp:287 at operator()(): DEBUG: No match due to default status for Package: pygments, Version: 2.11.2 while scanning for Vulnerability: CVE-2015-8557
2024/03/05 23:39:44 wazuh-modulesd:vulnerability-scanner[394838] packageScanner.hpp:287 at operator()(): DEBUG: No match due to default status for Package: pygments, Version: 2.11.2 while scanning for Vulnerability: CVE-2021-20270
2024/03/05 23:39:44 wazuh-modulesd:vulnerability-scanner[394838] packageScanner.hpp:287 at operator()(): DEBUG: No match due to default status for Package: pygments, Version: 2.11.2 while scanning for Vulnerability: CVE-2021-27291
2024/03/05 23:39:45 wazuh-modulesd:vulnerability-scanner[394838] packageScanner.hpp:287 at operator()(): DEBUG: No match due to default status for Package: pygments, Version: 2.16.1 while scanning for Vulnerability: CVE-2022-40896
``` 

- Affected
```
2024/03/05 23:39:44 wazuh-modulesd:vulnerability-scanner[394838] packageScanner.hpp:227 at operator()(): DEBUG: Match found, the package 'pygments', is vulnerable to 'CVE-2022-40896'. Current version: '2.11.2' (less than '2.15.0' or equal to ''). - Agent 'Thinkpad-Sebas' (ID: '000', Version: 'Wazuh v4.8.0').
```

In this scenario, the changes allowed the detection of another version of pygments, this seems related to:
- https://github.com/wazuh/wazuh/issues/22355

</details>

<details>
<summary><b>pip</b></summary>

### 4.8.0
- Unaffected (CVE-2013-1629, CVE-2013-1888, CVE-2013-5123, CVE-2014-8991)
```
2024/03/06 09:28:12 wazuh-modulesd:vulnerability-scanner[58809] packageScanner.hpp:271 at operator()(): DEBUG: No match due to default status for Package: pip, Version: 22.0.2 while scanning for Vulnerability: CVE-2013-1629
2024/03/06 09:28:12 wazuh-modulesd:vulnerability-scanner[58809] packageScanner.hpp:271 at operator()(): DEBUG: No match due to default status for Package: pip, Version: 22.0.2 while scanning for Vulnerability: CVE-2013-1888
2024/03/06 09:28:12 wazuh-modulesd:vulnerability-scanner[58809] packageScanner.hpp:271 at operator()(): DEBUG: No match due to default status for Package: pip, Version: 22.0.2 while scanning for Vulnerability: CVE-2013-5123
2024/03/06 09:28:12 wazuh-modulesd:vulnerability-scanner[58809] packageScanner.hpp:271 at operator()(): DEBUG: No match due to default status for Package: pip, Version: 22.0.2 while scanning for Vulnerability: CVE-2014-8991
```

- Failed to scan package (CVE-2019-20916, CVE-2021-3572, CVE-2023-5752)
```
2024/03/06 09:28:12 wazuh-modulesd:vulnerability-scanner[58809] packageScanner.hpp:125 at operator()(): DEBUG: Scanning package - 'pip' (Installed Version: 22.0.2, Security Vulnerability: CVE-2019-20916). Identified vulnerability: Version: 0. Required Version Threshold: 19.2. Required Version Threshold (or Equal): .
2024/03/06 09:28:12 wazuh-modulesd:vulnerability-scanner[58809] packageScanner.hpp:283 at operator()(): DEBUG: Failed to scan package: 'pip', CVE Numbering Authorities (CNA): 'pypi', Error: 'Unable to compare versions (22.0.2 vs 19.2).'
--
2024/03/06 09:28:12 wazuh-modulesd:vulnerability-scanner[58809] packageScanner.hpp:125 at operator()(): DEBUG: Scanning package - 'pip' (Installed Version: 22.0.2, Security Vulnerability: CVE-2021-3572). Identified vulnerability: Version: 0. Required Version Threshold: 21.1. Required Version Threshold (or Equal): .
2024/03/06 09:28:12 wazuh-modulesd:vulnerability-scanner[58809] packageScanner.hpp:283 at operator()(): DEBUG: Failed to scan package: 'pip', CVE Numbering Authorities (CNA): 'pypi', Error: 'Unable to compare versions (22.0.2 vs 21.1).'
--
2024/03/06 09:28:12 wazuh-modulesd:vulnerability-scanner[58809] packageScanner.hpp:125 at operator()(): DEBUG: Scanning package - 'pip' (Installed Version: 22.0.2, Security Vulnerability: CVE-2023-5752). Identified vulnerability: Version: 0. Required Version Threshold: 23.3. Required Version Threshold (or Equal): .
2024/03/06 09:28:12 wazuh-modulesd:vulnerability-scanner[58809] packageScanner.hpp:283 at operator()(): DEBUG: Failed to scan package: 'pip', CVE Numbering Authorities (CNA): 'pypi', Error: 'Unable to compare versions (22.0.2 vs 23.3).'
```

### New branch

- Unaffected (CVE-2013-1629, CVE-2013-1888, CVE-2013-5123, CVE-2014-8991, CVE-2019-20916, CVE-2021-3572)
```
2024/03/05 23:39:45 wazuh-modulesd:vulnerability-scanner[394838] packageScanner.hpp:287 at operator()(): DEBUG: No match due to default status for Package: pip, Version: 22.0.2 while scanning for Vulnerability: CVE-2013-1629
2024/03/05 23:39:45 wazuh-modulesd:vulnerability-scanner[394838] packageScanner.hpp:287 at operator()(): DEBUG: No match due to default status for Package: pip, Version: 22.0.2 while scanning for Vulnerability: CVE-2013-1888
2024/03/05 23:39:45 wazuh-modulesd:vulnerability-scanner[394838] packageScanner.hpp:287 at operator()(): DEBUG: No match due to default status for Package: pip, Version: 22.0.2 while scanning for Vulnerability: CVE-2013-5123
2024/03/05 23:39:45 wazuh-modulesd:vulnerability-scanner[394838] packageScanner.hpp:287 at operator()(): DEBUG: No match due to default status for Package: pip, Version: 22.0.2 while scanning for Vulnerability: CVE-2014-8991
2024/03/05 23:39:45 wazuh-modulesd:vulnerability-scanner[394838] packageScanner.hpp:287 at operator()(): DEBUG: No match due to default status for Package: pip, Version: 22.0.2 while scanning for Vulnerability: CVE-2019-20916
2024/03/05 23:39:45 wazuh-modulesd:vulnerability-scanner[394838] packageScanner.hpp:287 at operator()(): DEBUG: No match due to default status for Package: pip, Version: 22.0.2 while scanning for Vulnerability: CVE-2021-3572
```

- Affected (CVE-2023-5752)
```
2024/03/05 23:39:45 wazuh-modulesd:vulnerability-scanner[394838] packageScanner.hpp:227 at operator()(): DEBUG: Match found, the package 'pip', is vulnerable to 'CVE-2023-5752'. Current version: '22.0.2' (less than '23.3' or equal to ''). - Agent 'Thinkpad-Sebas' (ID: '000', Version: 'Wazuh v4.8.0')
```

The two packages that failed are now correctly identified as not affected:

- CVE-2019-20916
```json
{
    "defaultStatus": "unaffected",
    "product": "pip",
    "vendor": "pypi",
    "versions":
    [
        {
            "lessThan": "19.2",
            "status": "affected",
            "version": "0",
            "versionType": "custom"
        }
    ]
}
```

- CVE-2021-3572
```json
{
    "defaultStatus": "unaffected",
    "product": "pip",
    "vendor": "pypi",
    "versions":
    [
        {
            "lessThan": "19.2",
            "status": "affected",
            "version": "0",
            "versionType": "custom"
        }
    ]
}
```

The remaining one is correctly identified as affected:

- CVE-2023-5752
```json
{
    "product": "pip",
    "vendor": "pypi",
    "versions": [
        {
            "lessThan": "23.3",
            "status": "affected",
            "version": "0",
            "versionType": "custom"
        }
    ]
}
```

</details>

<details>
<summary><b>Setuptools</b></summary>

### 4.8.0
- Failed to scan package (CVE-2022-40897, CVE-2022-40897, CVE-2022-40897):
```
2024/03/06 09:27:56 wazuh-modulesd:vulnerability-scanner[58809] packageScanner.hpp:125 at operator()(): DEBUG: Scanning package - 'setuptools' (Installed Version: 56.0.0, Security Vulnerability: CVE-2022-40897). Identified vulnerability: Version: 0. Required Version Threshold: 65.5.1. Required Version Threshold (or Equal): .
2024/03/06 09:27:56 wazuh-modulesd:vulnerability-scanner[58809] packageScanner.hpp:283 at operator()(): DEBUG: Failed to scan package: 'setuptools', CVE Numbering Authorities (CNA): 'pypi', Error: 'Unable to compare versions (56.0.0 vs 65.5.1).'
--
2024/03/06 09:28:54 wazuh-modulesd:vulnerability-scanner[58809] packageScanner.hpp:125 at operator()(): DEBUG: Scanning package - 'setuptools' (Installed Version: 56.0.0, Security Vulnerability: CVE-2022-40897). Identified vulnerability: Version: 0. Required Version Threshold: 65.5.1. Required Version Threshold (or Equal): .
2024/03/06 09:28:54 wazuh-modulesd:vulnerability-scanner[58809] packageScanner.hpp:283 at operator()(): DEBUG: Failed to scan package: 'setuptools', CVE Numbering Authorities (CNA): 'pypi', Error: 'Unable to compare versions (56.0.0 vs 65.5.1).'
--
2024/03/06 09:31:48 wazuh-modulesd:vulnerability-scanner[58809] packageScanner.hpp:125 at operator()(): DEBUG: Scanning package - 'setuptools' (Installed Version: 59.6.0, Security Vulnerability: CVE-2022-40897). Identified vulnerability: Version: 0. Required Version Threshold: 65.5.1. Required Version Threshold (or Equal): .
2024/03/06 09:31:48 wazuh-modulesd:vulnerability-scanner[58809] packageScanner.hpp:283 at operator()(): DEBUG: Failed to scan package: 'setuptools', CVE Numbering Authorities (CNA): 'pypi', Error: 'Unable to compare versions (59.6.0 vs 65.5.1).'
```

### New branch
- Affected
```
2024/03/05 23:39:46 wazuh-modulesd:vulnerability-scanner[394838] packageScanner.hpp:227 at operator()(): DEBUG: Match found, the package 'setuptools', is vulnerable to 'CVE-2022-40897'. Current version: '59.6.0' (less than '65.5.1' or equal to ''). - Agent 'Thinkpad-Sebas' (ID: '000', Version: 'Wazuh v4.8.0').
```

The package that failed is now correctly detected as affected:

- CVE-2022-40897
```json
{
    "defaultStatus": "unaffected",
    "product": "setuptools",
    "vendor": "python",
    "versions":
    [
        {
            "lessThan": "65.5.1",
            "status": "affected",
            "version": "0",
            "versionType": "custom"
        }
    ]
}
```

</details>

### CODE REVIEW

**Review by MiguelazoDS — COMMENTED** [review]
GJ! just some findings 

I've got this errors with snapd package 

```console
2024/03/08 10:10:18 wazuh-modulesd:vulnerability-scanner[11989] packageScanner.hpp:299 at operator()(): DEBUG: Failed to scan package: 'snapd', CVE Numbering Authorities (CNA): 'canonical', Error: 'Unable to compare versions (2.61.2 vs 2.57.5+22.04ubuntu0.1).'
2024/03/08 10:10:18 wazuh-modulesd:vulnerability-scanner[11989] packageScanner.hpp:299 at operator()(): DEBUG: Failed to scan package: 'snapd', CVE Numbering Authorities (CNA): 'canonical', Error: 'Unable to compare versions (2.61.2 vs 2.58+22.04.1).'
```

```console
2024/03/08 10:10:18 wazuh-modulesd:vulnerability-scanner[11989] versionMatcher.hpp:146 at createVersionObject(): DEBUG: Error creating VersionObject (CalVer). Version string doesn't match the specified type. Version string: 2.61.2
2024/03/08 10:10:18 wazuh-modulesd:vulnerability-scanner[11989] versionMatcher.hpp:146 at createVersionObject(): DEBUG: Error creating VersionObject (CalVer). Version string doesn't match the specified type. Version string: 2.57.5+22.04ubuntu0.1
2024/03/08 10:10:18 wazuh-modulesd:vulnerability-scanner[11989] versionMatcher.hpp:157 at createVersionObject(): DEBUG: Error creating VersionObject (PEP440). Version string doesn't match the specified type. Version string: 2.57.5+22.04ubuntu0.1
2024/03/08 10:10:18 wazuh-modulesd:vulnerability-scanner[11989] versionMatcher.hpp:168 at createVersionObject(): DEBUG: Error creating VersionObject (MajorMinor). Version string doesn't match the specified type. Version string: 2.57.5+22.04ubuntu0.1
2024/03/08 10:10:18 wazuh-modulesd:vulnerability-scanner[11989] versionMatcher.hpp:146 at createVersionObject(): DEBUG: Error creating VersionObject (CalVer). Version string doesn't match the specified type. Version string: 2.61.2
2024/03/08 10:10:18 wazuh-modulesd:vulnerability-scanner[11989] versionMatcher.hpp:146 at createVersionObject(): DEBUG: Error creating VersionObject (CalVer). Version string doesn't match the specified type. Version string: 2.58+22.04.1
2024/03/08 10:10:18 wazuh-modulesd:vulnerability-scanner[11989] versionMatcher.hpp:157 at createVersionObject(): DEBUG: Error creating VersionObject (PEP440). Version string doesn't match the specified type. Version string: 2.58+22.04.1
2024/03/08 10:10:18 wazuh-modulesd:vulnerability-scanner[11989] versionMatcher.hpp:168 at createVersionObject(): DEBUG: Error creating VersionObject (MajorMinor). Version string doesn't match the specified type. Version string: 2.58+22.04.1
2024/03/08 10:10:18 wazuh-modulesd:vulnerability-scanner[11989] versionMatcher.hpp:179 at createVersionObject(): DEBUG: Error creating VersionObject (SemVer). Version string doesn't match the specified type. Version string: 2.58+22.04.1
```

This is the output of package inventory testtool 

image

One of them has deb format, I understand that for the other one is kinda expected since the strategy for snap is not implemented.

File: src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/packageScanner.hpp [hunk 1]
```diff
@@ -30,8 +31,23 @@ class TPackageScanner final : public AbstractHandler<std::shared_ptr<TScanContex
...
+     * name.
+     */
+    std::unordered_map<std::string_view, std::variant<VersionObjectType, VersionMatcherStrategy>> m_packageMap {
+        {"deb", VersionObjectType::DPKG},
+        {"rpm", VersionObjectType::RPM},
+        {"pypi", VersionObjectType::PEP440},
+        {"npm", VersionObjectType::SemVer},
+        {"pacman", VersionMatcherStrategy::Pacman},
+        {"snap", VersionMatcherStrategy::Snap},
+        {"pkg", VersionMatcherStrategy::PKG},
```
[the remark this answers is not in the document]
Comment (sebasfalcone) [reply src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/packageScanner.hpp:48]: Given that both Solaris and MacOS make use of pkg I think it should have its own strategy (pending investigation to determine it)

File: src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/versionMatcher/versionMatcher.hpp [hunk 2]
```diff
@@ -40,145 +53,190 @@ class VersionMatcher final
...
+     */
+    static std::shared_ptr<IVersionObject> createVersionObject(const std::string& version,
+                                                               const VersionMatcherStrategy& strategy)
+    {
+        std::shared_ptr<IVersionObject> matcher;
+
+        switch (strategy)
+        {
+            case VersionMatcherStrategy::Windows:
+            case VersionMatcherStrategy::MacOS: return createVersionObject(version, VersionObjectType::DPKG); break;
```
[the remark this answers is not in the document]
Comment (sebasfalcone) [reply src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/versionMatcher/versionMatcher.hpp:76]: This was proposed by @Dwordcito and seems to be the best option

File: src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/packageScanner.hpp [hunk 3]
```diff
@@ -30,8 +31,23 @@ class TPackageScanner final : public AbstractHandler<std::shared_ptr<TScanContex
...
 private:
     std::shared_ptr<TDatabaseFeedManager> m_databaseFeedManager;
-    std::unordered_map<std::string_view, VersionObjectType> m_mapVendors {{"deb", VersionObjectType::DPKG},
-                                                                          {"rpm", VersionObjectType::RPM}};
+
+    /**
+     * @brief Package format to VersionObjectType / VersionMatcherStrategy map.
+     *
+     * @note The map is used to determine the version object type or the version matcher strategy based on the package
+     * name.
```
[the remark this answers is not in the document]
Comment (sebasfalcone) [reply src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/packageScanner.hpp:39]:
```suggestion
     * format.
```

File: src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/packageScanner.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (sebasfalcone) [reply src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/packageScanner.hpp:48]: I've added the apk format to the map

**Review by MiguelazoDS — APPROVED** [review]

**Comment by sebasfalcone** [comment]
Failing check related to:
- https://github.com/wazuh/wazuh/issues/22424

File: src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/versionMatcher/versionMatcher.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (MiguelazoDS) [reply src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/versionMatcher/versionMatcher.hpp:77]: This is matching macports, I see now.

**Pull request merged**